File: 0567-permutation-in-string/0567-permutation-in-string.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        if(s1.size() > s2.size()) return false;
        
        vector<int> charFreq(26, 0);
        vector<int> tempCharFreq(26, 0);
        
        for(int i = 0; i < s1.size(); i++) {
            charFreq[s1[i] -'a']++;
            tempCharFreq[s2[i] -'a']++;
        }
        
        if(charFreq == tempCharFreq) return true;
        
        for(int k = 0, i = s1.size(); i < s2.size(); i++, k++) {
            tempCharFreq[s2[k] -'a']--;
            tempCharFreq[s2[i] -'a']++;
            
            if(charFreq == tempCharFreq) return true;
        }
        
        return false;
    }
};

File: 0567-permutation-in-string/0567-permutation-in-string.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        final int M = s1.length();
        final int N = s2.length();
        
        if(M > N) return false;
        
        int[] charFreq1 = new int[26];
        int[] charFreq2 = new int[26];
        
        for(int i = 0; i < M; i++) {
            int idx1 = s1.charAt(i) -'a';
            int idx2 = s2.charAt(i) -'a';
            
            charFreq1[idx1]++;
            charFreq2[idx2]++;
        }
        
        int matchChars = 0;
        
        for(int i = 0; i < 26; i++) {
            if(charFreq1[i] == charFreq2[i]) matchChars++;
        }
        
        int i = 0, j = M;
        while(j < N) {
            int left = s2.charAt(i) -'a';
            int right = s2.charAt(j) -'a';
            
            if(matchChars == 26) {
                return true;
            }
            
            charFreq2[right]++;
            if(charFreq1[right] == charFreq2[right]) {
                matchChars++;
            } else if(charFreq1[right] == charFreq2[right] -1) {
                matchChars--;
            }
            
            charFreq2[left]--;
            if(charFreq1[left] == charFreq2[left]) {
                matchChars++;
            } else if(charFreq1[left] == charFreq2[left] +1) {
                matchChars--;
            }
            
            i++;
            j++;
        }
        
        return matchChars == 26;
    }
}